pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go

```diff
@@ -0,0 +1,506 @@
+		Description:         fmt.Sprintf(`{"kubernetes.io/service-name":"%s"}`, nm.String()),
+		IPAddress:           svc.Spec.LoadBalancerIP,
+		BackendService:      backendServiceLink,
+		Ports:               ports,
+		IPProtocol:          string(protocol),
+		LoadBalancingScheme: string(scheme),
+	}
+
+	// Specify subnetwork if network type is manual
+	if len(gce.subnetworkURL) > 0 {
+		wantedFwdRule.Subnetwork = gce.subnetworkURL
+	} else {
+		wantedFwdRule.Network = gce.networkURL
+	}
+
+	// Delete forwarding rule if any settings are different
+	fwdRuleDeleted := false
+	if existingFwdRule != nil && !fwdRuleEqual(existingFwdRule, wantedFwdRule) {
+		glog.V(2).Infof("ensureInternalLoadBalancer(%v(%v)): deleting existing forwarding rule with IP address %v", loadBalancerName, svc.Name, existingFwdRule.IPAddress)
+		if err := gce.DeleteRegionForwardingRule(loadBalancerName, gce.region); err != nil {
+			return nil, err
+		}
+		fwdRuleDeleted = true
+	}
+
+	// Ensure instance groups exist and nodes are assigned to groups
+	igName := getInstanceGroupName(clusterID)
+	igLinks, err := gce.ensureInstanceGroups(igName, nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get health check settings assuming it's an OnlyLocal service
+	name := getHealthCheckName(loadBalancerName, clusterID, shared)
+	hcPath, hcPort := GetNodesHealthCheckPath(), GetNodesHealthCheckPort()
+	if !shared {
+		hcPath, hcPort = v1_service.GetServiceHealthCheckPathPort(svc)
+	}
+	hc, err := gce.ensureHealthCheck(name, hcPath, hcPort)
+	if err != nil {
+		return nil, err
+	}
+
+	balancingMode := balanceModeRATE
+	if scheme == schemeInternal {
+		balancingMode = balanceModeCONNECTION
+	}
+
+	err = gce.ensureBackendService(backendServiceName, scheme, balancingMode, protocol, igLinks, hc.SelfLink)
+	if err != nil {
+		return nil, err
+	}
+
+	if fwdRuleDeleted || existingFwdRule == nil {
+		if err = gce.CreateRegionForwardingRule(wantedFwdRule, gce.region); err != nil {
+			return nil, err
+		}
+	}
+
+	if !gce.OnXPN() {
+		if err = gce.ensureFirewalls(loadBalancerName, clusterID, nm, svc, strconv.Itoa(int(hcPort)), shared, nodes); err != nil {
+			return nil, err
+		}
+	} else {
+		glog.V(2).Infof("ensureInternalLoadBalancer: cluster is on a cross-project network (XPN) network project %v, compute project %v - skipping firewall creation", gce.networkProjectID, gce.projectID)
+	}
+
+	existingFwdRule, err = gce.GetRegionForwardingRule(loadBalancerName, gce.region)
+	if err != nil {
+		return nil, err
+	}
+
+	status := &v1.LoadBalancerStatus{}
+	status.Ingress = []v1.LoadBalancerIngress{{IP: existingFwdRule.IPAddress}}
+	return status, nil
+}
+
+func (gce *GCECloud) updateInternalLoadBalancer(clusterName, clusterID string, service *v1.Service, nodes []*v1.Node) error {
+	igName := getInstanceGroupName(clusterID)
+	_, err := gce.ensureInstanceGroups(igName, nodes)
+	return err
+}
+
+func (gce *GCECloud) ensureInternalLoadBalancerDeleted(clusterName, clusterID string, svc *v1.Service) error {
```

Comment:
Didn't see the sharedResourceLock is used, did it turn out that ILB management does not need locking?